### ISSUE TITLE
1.3.2 packer amazon-ebs builder process unresponsive when credentials aren't present (build / validate)

### ISSUE THREAD

**Issue by eedwards-sk** [hashicorp/packer#6888]
# with

1. given the following template as `template.json`:

```
{
    "min_packer_version": "1.3.1",
    "variables": {
        "aws_region": null
    },
    "builders": [
        {
            "name": "ubuntu18-ami",
            "ami_name": "concourse-packer-test-{{isotime | clean_ami_name}}",
            "ami_description": "an ubuntu 18.04 ami built with packer using concourse",
            "instance_type": "t2.micro",
            "region": "{{user `aws_region`}}",
            "type": "amazon-ebs",
            "source_ami_filter": {
                "filters": {
                    "virtualization-type": "hvm",
                    "architecture": "x86_64",
                    "name": "ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*",
                    "block-device-mapping.volume-type": "gp2",
                    "root-device-type": "ebs"
                },
                "owners": [
                    "099720109477"
                ],
                "most_recent": true
            },
            "ssh_username": "ubuntu"
        }
    ],
    "provisioners": [
        {
            "type": "shell",
            "inline": [
                "apt-get update -y"
            ]
        }
    ]
}
```

2. given an environment which does not have any aws credentials available

3. given the command `packer build -var=aws_region=us-east-1 template.json`

# expected results:

build command proceeds, likely with an error related to credentials missing, and then the process exits

# actual results

```
→ packer build -var=aws_region=us-east-1 template.json
ubuntu18-ami output will be in this color.
```

this is the only output, and you cannot `Ctrl+C` the process, I have to `Ctrl+Z` it (background it), and then do a `kill -9` on the process... it won't respond at all. completely frozen / hung.

I can repro this with packer installed on OS X Mojave as well as inside an alpine docker container, so I suspect it's a problem with the binary/version, not the platform.

**Comment by SwampDragons** [comment]
This is a problem in v 1.3.1 but has already been fixed on master and will be released with v 1.3.2

**Comment by eedwards-sk** [comment]
@SwampDragons 

Please re-open this.

It's still broken in 1.3.2, and now it's worse :(

Given the same above template, running in docker containers one with `1.3.2` and one with `1.3.1`, build still hangs in both versions, but now `validate` also hangs in `1.3.2`.

given validate command:

```
PACKER_LOG=1 packer validate -machine-readable -var=aws_region=us-east-1 test/data/template.json
```

See this [gist](https://gist.github.com/eedwards-sk/fafddb44a0a20b312c28aa5606d16989) showing output for  where it works on 1.3.1 but hangs on 1.3.2 (requiring a total of four `Ctrl+C` interrupts to exit, and leaving a hung process `/bin/packer plugin packer-builder-amazon-ebs`)

In addition, `packer build` still hangs on `1.3.2` with no indication of credential issues, however it does exit with `Ctrl+C` whereas `1.3.1` still requires a `Ctrl+Z` and then `kill -9` the job

**Comment by eedwards-sk** [comment]
Ahh, for clarity, I changed my template since the above.

I did add some aws variables. The updated template is below:

```
{
    "min_packer_version": "1.3.1",
    "variables": {
        "aws_region": null,
        "aws_access_key_id": "",
        "aws_secret_access_key": ""
    },
    "sensitive-variables": [
        "aws_access_key_id",
        "aws_secret_access_key"
    ],
    "builders": [
        {
            "access_key": "{{user `aws_access_key_id`}}",
            "secret_key": "{{user `aws_secret_access_key`}}",
            "name": "ubuntu18-ami",
            "ami_name": "concourse-packer-test-{{isotime | clean_ami_name}}",
            "ami_description": "an ubuntu 18.04 ami built with packer using concourse",
            "instance_type": "t2.micro",
            "region": "{{user `aws_region`}}",
            "type": "amazon-ebs",
            "source_ami_filter": {
                "filters": {
                    "virtualization-type": "hvm",
                    "architecture": "x86_64",
                    "name": "ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*",
                    "block-device-mapping.volume-type": "gp2",
                    "root-device-type": "ebs"
                },
                "owners": [
                    "099720109477"
                ],
                "most_recent": true
            },
            "ssh_username": "ubuntu"
        }
    ],
    "provisioners": [
        {
            "type": "shell",
            "inline": [
                "sudo apt-get update -y"
            ]
        }
    ]
}
```

However, this still raises the question of why it's touching aws at all on a validate?

Also, still no error output indicating it's related to credentials, just a hung process.

**Comment by SwampDragons** [comment]
> Also, I'm concerned why it would even be trying to touch aws credentials on a validate.

I can answer that for you. We changed the region validation to actually query AWS for a list of regions, rather than depending on a hardcoded list.  The idea was to prevent Packer from inflexibly requiring a person to update code, but it looks like there are situations where people really need Packer to be able to validate without using credentials.  

For now, you can work around this issue by using the [skip-region-validation](https://www.packer.io/docs/builders/amazon-ebs.html#skip_region_validation) template option, but obviously this is an unintended consequence that's probably worse than having had the hardcoded list.

I'm going to think about the best way to resolve this.

**Comment by eedwards-sk** [comment]
@SwampDragons Thanks for the update and the workaround.

For now I'm trying `-syntax-only` which seems to also skip a lot of the validation.

Honestly I don't mind it possibly needing credentials for validation when you're at the build phase (where you expect to have all variables at hand), but for unit testing or simple PR checking etc, ideally as little real configuration or credentials would be needed (and maybe that's what the `-syntax-only` option is best suited for).

Either way, having the process die gracefully with error output when missing credentials would be ideal ;)

**Comment by SwampDragons** [comment]
>Either way, having the process die gracefully with error output when missing credentials would be ideal ;)

Yeah, it would be and I'm not sure why amazon's bindings _aren't_ doing that.

**Comment by SwampDragons** [comment]
Some thoughts on validation options, just for conversation and posterity.

- We could use a separate hardcoded list for validation, and just accept that it won't be totally up to date.
	- Pros: Simple, easy to implement, and fixes validation
	- Cons: Loses some of the actual point of validation.

- We could generate the list upon `make`. 
	Pros: This won't be cutting-edge up-to-date but it would be fairly complete
	Cons: 
		- Would only populate with items that the current user has, which makes things exciting when thinking about the various Government regions, not to mention Chinese ones.
		- Requires all people building Packer to have AWS creds.

- We could allow users to provide a validation list?
	Pros: People get what they want.
	Cons: 
		- Extra configuration when people want this to "Just Work"
		- Essentially renders validation meaningless. Just use `skip_region_validation`.

- We could revert back to what we were doing and use a hardcoded list for everything
	- Pros: Simple, we know it works for most cases.
	- Cons: Back to the same old issue of needing to update Packer manually every time AWS does an update.

- Hybrid approach: We could use the hardcoded list but try to fall back on using a real query if the region does not exist in the hardcoded list.
	- Pros: Probably best of both worlds. Get the speed and reliability of validation via a hardcoded list but the flexibiltiy of people being able to continue using older builds of Packer even as new regions come out.
	- Cons: Just obfuscates this issue with credentials and pushes it into a smaller edge case.

**Comment by eedwards-sk** [comment]
Out of curiosity, may I ask why packer believes it needs to be an authority on the region value?

I assume there is an underlying aws api that it's leveraging when building, and that providing an invalid value there should surface an error in some fashion (I would hope).

As an operator / developer / administrator, I expect that if I provide an invalid value for an aws region, it would be surfaced at **build time** which is when I expect credentials to be needed and exercised, as they are then provided to the upstream API.

Prior to build time, I expect any validation to be a 'dry run', and not surface an invalid region.

However, that doesn't seem to be the model packer uses.

So from my perspective, If packer indeed intends to exercise credentials at validation time, then live region validation makes sense, as you always need credentials, so what's the value of packer having any hardcoded concept of valid or invalid regions itself?

An option is to have a clear boundary between 'exercising' and 'not exercising' credentials (maybe that is `-syntax-only` or `skip_region_validation`? clear nomenclature would help too), then when doing the 'dry run' (not exercising credentials), I would want and expect packer to **not** do any region validation against any pre-known regions.

Put simply, from my perspective, if packer wants to be "smart" about regions, it should require credentials and be explicit about it, and pass validation to the upstream api (get actual list of regions).

If instead we can let it be "dumb", ideally surface the boundary clearly, and don't get in the way of region validation at all, with the operator assuming that only takes place when explicitly validating 'live', or during a build.

Hope that makes sense. Just some musings from my experience trying to get clean boundaries between validating for syntax/configuration and trying actual builds. I want to avoid as much 'environment binding' (inputting real credentials or values) at the dev/unit phase, as possible.

**Comment by SwampDragons** [comment]
Hmm. You're probably right that `packer validate` doesn't _really_ need to tell us whether the region is real. I remember being conflicted about allowing this API call inside of Prepare() when the original PR to change this code was made, and now I'm regretting letting it through. 

It is important to do during the Prepare() portion of the build itself, though, because we'd rather not have Packer spend hours running and then fail last-minute because someone wrote "us-est-1" instead of "us-east-1".

The problem is that `packer validate` works by calling the builder's Prepare() method and then just not calling the Run() part of the builder.  So allowing this API call into the build's Prepare() broke things.

Maybe the right answer to the validate/build divide may be to just move the region validation into the beginning of the Run() portion of the build instead so that it's still called pre-flight but not during the `packer validate` call. It's a small decrease in functionality but may be worth it to remove this lack of clarity.

**Comment by eedwards-sk** [comment]
>because we'd rather not have Packer spend hours running and then fail last-minute

Great point! I agree that having the failure at the end of a long run would suck, so I appreciate the intent of the original change.

>Maybe the right answer to the validate/build divide may be to just move the region validation into the beginning of the Run() portion of the build instead so that it's still called pre-flight but not during the packer validate call. It's a small decrease in functionality but may be worth it to remove this lack of clarity.

To me, at least from an outsider's perspective, that sounds like the most rational approach. If I understand correctly that would offer the ability to fail early when trying to build with "us-derp-1", while also eliminating the burden of supplying real values during `validate`.

**Comment by SwampDragons** [comment]
Yeah I can see that. It feels weird to do, just because "we've always done things this way" -- which isn't a great reason to stick with things. I'll do a little spelunking through the history to see if I can suss out the original intent and whether this violates that intent.

**Comment by azr** [comment]
Okay, I just made a PR https://github.com/aws/aws-sdk-go/pull/2298 which should fix the empty credentials issue.
